src/metabase/api/dashboard.clj

```diff
@@ -607,6 +564,93 @@
+           (when (not (false? (:archived false)))
+             (api/check-not-archived dash-before-update))
+           (let [{current-dashcards :dashcards
+                  current-tabs      :tabs}  (t2/hydrate dash-before-update [:dashcards :series :card] :tabs)
+                 new-tabs                   (map-indexed (fn [idx tab] (assoc tab :position idx)) tabs)
+                 {:keys [old->new-tab-id
+                         deleted-tab-ids]
+                  :as   tabs-changes-stats} (dashboard-tab/do-update-tabs! (:id dash-before-update) current-tabs new-tabs)
+                 deleted-tab-ids            (set deleted-tab-ids)
+                 current-dashcards          (remove (fn [dashcard]
+                                                      (contains? deleted-tab-ids (:dashboard_tab_id dashcard)))
+                                                    current-dashcards)
+                 new-dashcards              (cond->> dashcards
+                                              ;; fixup the temporary tab ids with the real ones
+                                              (seq old->new-tab-id)
+                                              (map (fn [card]
+                                                     (if-let [real-tab-id (get old->new-tab-id (:dashboard_tab_id card))]
+                                                       (assoc card :dashboard_tab_id real-tab-id)
+                                                       card))))
+                 dashcards-changes-stats    (do-update-dashcards! dash-before-update current-dashcards new-dashcards)]
+             (reset! changes-stats
+                     (merge
+                      (select-keys tabs-changes-stats [:created-tab-ids :updated-tab-ids :deleted-tab-ids :total-num-tabs])
+                      (select-keys dashcards-changes-stats [:created-dashcards :deleted-dashcards :updated-dashcards]))))))
+       true))
+    (let [dashboard (t2/select-one :model/Dashboard id)]
+      (when update-dashboard-itself?
+        ;; execute these events for old PUT /api/dashboard/:id calls and new PUT /api/dashboard/:id calls, and not for old PUT/api/dashboard/:id/card calls
+        (events/publish-event! :event/dashboard-update {:object dashboard :user-id api/*current-user-id*}))
+      (when update-dashcards-and-tabs?
+        ;; execute these events for old PUT /api/dashboard/:id/cards calls and new PUT /api/dashboard/:id calls, and not for old PUT /api/dashboard/:id calls
+        (track-dashcard-and-tab-events! dashboard @changes-stats))
+      (-> (t2/hydrate dashboard [:collection :is_personal] [:dashcards :series] :tabs)
+          (assoc :last-edit-info (last-edit/edit-information-for-user @api/*current-user*))))))
+
+(api/defendpoint PUT "/:id"
+  "Update a Dashboard, and optionally the `dashcards` and `tabs` of a Dashboard. The request body should be a JSON object with the same
+  structure as the response from `GET /api/dashboard/:id`."
```

Comment:
my 2 cents: ideally PUT endpoint response should have the same structure as GET endpoint response. Is that the case?

Reply:
That's not the case, and wasn't before this PR. The goal of this series of PRs is to preserve the current behaviour and make the dashboard save much faster. I would say that making the PUT endpoint return the same as the GET endpoint is desirable and we can do that too

Reply:
FYI before this PR `tabs` were not returned from PUT endpoint. This caused an issue here: https://github.com/metabase/metabase/pull/35311#issuecomment-1792115937

Reply:
Wow, thanks for showing me this. I've been expecting a bug to happen like that for a while, but I was told not to worry about it. It's about time we do something about it! Open to ideas. I think it makes sense to have a shared schema for the FE + BE that both sides know how to edit.